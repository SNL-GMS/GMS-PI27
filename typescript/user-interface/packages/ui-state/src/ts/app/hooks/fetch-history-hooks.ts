import type { SerializedAxiosError } from '@gms/ui-workers';
import type { SerializedError } from '@reduxjs/toolkit';
import flatMap from 'lodash/flatMap';
import React from 'react';

import type { AsyncFetchHistory, AsyncFetchHistoryEntry, FetchHistoryStatus } from '../query';
import { AsyncActionStatus } from '../query';

/**
 * A hook that can be used to return the current status of retrieving the async fetch history.
 * This includes the following information:
 *  - the async fetch status of all the async requests
 *  - the `data`: the history of the query
 *
 * @returns returns the current status of retrieving a query.
 */
export function useFetchHistoryStatus<T = unknown>(
  history: AsyncFetchHistory<T>
): FetchHistoryStatus {
  const [pending, fulfilled, rejected, isLoading, isError, error] = React.useMemo(() => {
    const statuses: AsyncFetchHistoryEntry<T>[] = flatMap(
      Object.values(history).map(entry => Object.values(entry))
    );

    const status = statuses.reduce<{
      pending: number;
      fulfilled: number;
      rejected: number;
      error: (SerializedError | SerializedAxiosError)[];
    }>(
      (prev, current) => ({
        pending: current.status === AsyncActionStatus.pending ? prev.pending + 1 : prev.pending,
        fulfilled:
          current.status === AsyncActionStatus.fulfilled ? prev.fulfilled + 1 : prev.fulfilled,
        rejected: current.status === AsyncActionStatus.rejected ? prev.rejected + 1 : prev.rejected,
        error: current.error ? prev.error.concat(current.error) : prev.error
      }),
      { pending: 0, fulfilled: 0, rejected: 0, error: [] }
    );

    return [
      status.pending,
      status.fulfilled,
      status.rejected,
      status.pending > 0,
      // flag an error if there at least one rejected request and no pending/fulfilled requests
      status.pending < 1 && status.fulfilled < 1 && status.rejected > 0,
      status.error
    ];
  }, [history]);

  return React.useMemo(() => {
    return {
      pending,
      fulfilled,
      rejected,
      isLoading,
      isError,
      error: isError ? error : undefined
    };
  }, [error, fulfilled, isError, isLoading, pending, rejected]);
}
