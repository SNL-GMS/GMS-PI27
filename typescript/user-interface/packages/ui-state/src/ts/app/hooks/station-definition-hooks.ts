import { FacetedTypes, type StationTypes } from '@gms/common-model';
import flatMap from 'lodash/flatMap';
import sortBy from 'lodash/sortBy';
import uniq from 'lodash/uniq';
import React from 'react';

import { useGetProcessingStationGroupNamesConfigurationQuery } from '../api/processing-configuration';
import type { StationQuery } from '../api/station-definition';
import {
  useGetStationGroupsByNamesQuery,
  useGetStationsQuery,
  useGetStationsWithChannelsQuery
} from '../api/station-definition';
import { useEffectiveTime } from './operational-time-period-configuration-hooks';
import { useOldQueryDataIfReloading } from './query-util-hooks';
import { useAppSelector } from './react-redux-hooks';
import { useGetVisibleStationsFromStationList } from './waveform-hooks';

/**
 * Queries for provided effective time and station group names, query for all stations.
 *
 * Uses the `useGetStationsQuery` query hook.
 *
 * @returns returns the station query result
 */
const useGetStationGroupsAndGetStationQuery = (
  effectiveTime: number,
  stationGroupNames: string[]
): StationQuery => {
  const stationGroupsByNamesQuery = useGetStationGroupsByNamesQuery({
    effectiveTime,
    stationGroupNames
  });

  // If there are station groups get unique sorted list of station names
  const stationNames = React.useMemo(() => {
    const stationGroups = stationGroupsByNamesQuery.data;
    return sortBy(uniq(flatMap((stationGroups || []).map(x => x.stations.map(y => y.name)))));
  }, [stationGroupsByNamesQuery.data]);

  return useGetStationsQuery({ effectiveTime, stationNames });
};

/**
 * Queries for provided effective time, query for all stations for the configured station groups.
 *
 * Uses the `useGetStationsQuery` query hook.
 *
 * @returns returns the station query result
 */
export const useGetAllStationsQuery = (effectiveTime: number): StationQuery => {
  const processingStationGroupNamesConfiguration =
    useGetProcessingStationGroupNamesConfigurationQuery();
  const stationGroupNames = React.useMemo(
    () => sortBy(processingStationGroupNamesConfiguration.data?.stationGroupNames ?? []),
    [processingStationGroupNamesConfiguration.data?.stationGroupNames]
  );
  return useGetStationGroupsAndGetStationQuery(effectiveTime, stationGroupNames);
};

/**
 * Queries for provided effective time, query for all stations for the configured station groups.
 *
 * Uses the `useGetStationsQuery` query hook.
 *
 * @returns returns the station query result
 */
export const useGetAllStationsWithChannelsQuery = (effectiveTime: number): StationQuery => {
  const processingStationGroupNamesConfiguration =
    useGetProcessingStationGroupNamesConfigurationQuery();
  const stationGroupNames = React.useMemo(
    () => sortBy(processingStationGroupNamesConfiguration.data?.stationGroupNames ?? []),
    [processingStationGroupNamesConfiguration.data?.stationGroupNames]
  );
  const stationGroupsByNamesQuery = useGetStationGroupsByNamesQuery({
    effectiveTime,
    stationGroupNames
  });
  const stationNames = React.useMemo(
    () =>
      sortBy(uniq(flatMap(stationGroupsByNamesQuery.data?.map(x => x.stations.map(y => y.name))))),
    [stationGroupsByNamesQuery.data]
  );
  return useGetStationsWithChannelsQuery({ effectiveTime, stationNames });
};

/**
 * Uses all stations query and visible stations list hooks to return an array of stations visible in the waveform display
 *
 * @returns return an array of visible stations
 */
export const useVisibleStations = (): StationTypes.Station[] => {
  const effectiveAt = useEffectiveTime();
  const stationsQuery = useGetAllStationsQuery(effectiveAt);
  const getVisibleStationsFromStationList = useGetVisibleStationsFromStationList();
  const stationData = useOldQueryDataIfReloading<StationTypes.Station[]>(stationsQuery);
  return React.useMemo(
    () => (stationData ? getVisibleStationsFromStationList(stationData) : []),
    [getVisibleStationsFromStationList, stationData]
  );
};

/**
 *
 * @returns an array of version references of visible stations
 */
export const useGetVisibleStationsVersionReferences = (): FacetedTypes.VersionReference<
  'name',
  StationTypes.Station
>[] => {
  const visibleStations = useVisibleStations();
  return React.useMemo(() => {
    if (visibleStations) {
      return visibleStations.map(station =>
        FacetedTypes.convertToVersionReference(
          { effectiveAt: station.effectiveAt, name: station.name },
          'name'
        )
      );
    }
    return [];
  }, [visibleStations]);
};

/**
 * Uses all stations query to return an array of all stations with channels for this effective time
 *
 * @returns return an array of all stations
 */
export const useAllStations = (): StationTypes.Station[] => {
  const empty = React.useRef<StationTypes.Station[]>([]);
  const effectiveAt = useEffectiveTime();
  const stationsQuery = useGetAllStationsQuery(effectiveAt);
  return useOldQueryDataIfReloading<StationTypes.Station[]>(stationsQuery) || empty.current;
};

/**
 * Create a new station group with the effective time from the workflow redux state start time and the current station group.
 * Used for fetching processing masks.
 *
 * @returns a station group matching the one currently open via workflow, effective at the workflow start time
 */
export function useStationGroupEffectiveForInterval() {
  const stationGroup = useAppSelector(state => state.app.workflow.stationGroup);
  const startTimeSecs = useAppSelector(state => state.app.workflow.timeRange.startTimeSecs);

  return React.useMemo(
    () => ({
      ...stationGroup,
      effectiveAt: startTimeSecs || stationGroup.effectiveAt
    }),
    [startTimeSecs, stationGroup]
  );
}
