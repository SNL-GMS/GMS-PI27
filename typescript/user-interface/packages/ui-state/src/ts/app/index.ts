export * from './api';
export * from './history';
export * from './hooks';
export * from './loader';
export * from './processing';
export * from './query';
export * from './redux-provider';
export * from './sanitization';
export * from './state';
export * from './store';
export * from './subscription';
export * from './util';
