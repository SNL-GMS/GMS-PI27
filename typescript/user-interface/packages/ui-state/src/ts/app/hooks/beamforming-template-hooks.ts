import { BeamformingTemplateTypes } from '@gms/common-model';
import { unwrapResult } from '@reduxjs/toolkit';
import React from 'react';

import type { GetBeamformingTemplatesQueryArgs } from '../api/data/signal-enhancement';
import { getBeamformingTemplates } from '../api/data/signal-enhancement';
import { UIStateError } from '../error-handling/ui-state-error';
import type { AsyncFetchResult, FetchHistoryStatus } from '../query';
import { handleCanceledRequests } from '../query/async-fetch-util';
import { useFetchHistoryStatus } from './fetch-history-hooks';
import { useAppDispatch, useAppSelector } from './react-redux-hooks';
import { usePhaseTypeFavorites } from './signal-detection-hooks';
import { useGetVisibleStationsVersionReferences } from './station-definition-hooks';

/**
 * Defines async fetch result for the beamforming templates. It contains flags indicating
 * the status of the request.
 *
 * @see {@link AsyncFetchResult}
 */
export type BeamformingTemplateFetchResult =
  AsyncFetchResult<BeamformingTemplateTypes.BeamformingTemplatesByStationByPhase>;

/**
 * A hook that can be used to return the current history of the beamforming templates query.
 * This includes the following information:
 *  - the async fetch status of all the async requests
 *  - the `data`: the history of the `getBeamformingTemplates` queries
 *
 * @returns the current history of the processing mask definitions query.
 */
export const useGetBeamformingTemplatesRangeHistory = (): FetchHistoryStatus => {
  const history = useAppSelector(state => state.data.queries.getBeamformingTemplates);
  return useFetchHistoryStatus<GetBeamformingTemplatesQueryArgs>(history);
};

/**
 * A hook that issues the requests for the beamforming templates query.
 *
 * @param args the beamforming templates query arguments
 */
const useFetchBeamformingTemplatesQuery = (args: GetBeamformingTemplatesQueryArgs): void => {
  const dispatch = useAppDispatch();
  //! useEffect updates redux state
  React.useEffect(() => {
    dispatch(getBeamformingTemplates(args)).catch(error => {
      throw new UIStateError(error);
    });
  }, [dispatch, args]);
};

/**
 * Check if all stations and phases in the request args are represented in the
 * provided data object (which is presumably coming from Redux)
 *
 * @param data The redux record of records representing all rotation templates that we have fetched
 * @param args the args for a possible query
 * @returns true if the rotation templates already exist in data that would be fetched for query represented by args
 */
export function hasBeamformingTemplatesForQueryArgs(
  data: BeamformingTemplateTypes.BeamformingTemplatesByBeamTypeByStationByPhase,
  args: GetBeamformingTemplatesQueryArgs
): boolean {
  const beamformingTemplatesByStationByPhase = data[args.beamType];
  if (!beamformingTemplatesByStationByPhase) {
    return false;
  }
  return args.stations.every(station => {
    const beamformingTemplatesByPhase = beamformingTemplatesByStationByPhase[station.name];
    if (!beamformingTemplatesByPhase) {
      return false;
    }
    return args.phases.every(phase => beamformingTemplatesByPhase[phase] != null);
  });
}

/**
 * Hook to create a callback for getting/fetching beamforming templates
 *
 * @returns a callback function that returns {@link BeamformingTemplatesByStationByPhase} for the {@link BeamformingTemplateTypes.BeamType} containing the requested PhasesAndStations.
 * If these already exist in redux it will return the full object from redux which may contain additional stations and phases.
 * Otherwise it will fetch the beam templates and return the result
 */
export const useFetchBeamformingTemplatesQueryFunction = () => {
  const dispatch = useAppDispatch();
  const data = useAppSelector(state => state.data.beamformingTemplates);

  return React.useCallback(
    async (
      args: GetBeamformingTemplatesQueryArgs
    ): Promise<BeamformingTemplateTypes.BeamformingTemplatesByStationByPhase> => {
      if (hasBeamformingTemplatesForQueryArgs(data, args)) {
        return data[args.beamType];
      }
      return dispatch(getBeamformingTemplates(args))
        .then(handleCanceledRequests(unwrapResult))
        .catch(error => {
          throw new UIStateError(error.message);
        });
    },
    [data, dispatch]
  );
};

/**
 * A hook that can be used to retrieve beamforming templates
 *
 * @returns the beamforming templates result.
 */
export const useBeamformingTemplates = (
  args: GetBeamformingTemplatesQueryArgs
): BeamformingTemplateFetchResult => {
  const history = useGetBeamformingTemplatesRangeHistory();
  // issue any new fetch requests
  useFetchBeamformingTemplatesQuery(args);

  // retrieve all beamforming templates from the state
  const beamformingTemplates = useAppSelector(state => state.data.beamformingTemplates);
  return React.useMemo(() => {
    return { ...history, data: beamformingTemplates[args.beamType] };
  }, [args, history, beamformingTemplates]);
};

/**
 * Hook to obtain beamforming templates for visible stations and favorite phases
 *
 * @param beamType
 * @returns beamFormingTemplate
 */
export const useBeamformingTemplatesForVisibleStationsAndFavoritePhases = (
  beamType: BeamformingTemplateTypes.BeamType,
  additionalPhases?: string[]
): BeamformingTemplateFetchResult => {
  const stationVersionReferences = useGetVisibleStationsVersionReferences();
  const phases = usePhaseTypeFavorites(additionalPhases);

  const args = React.useMemo(
    () => ({
      phases,
      stations: stationVersionReferences,
      beamType
    }),
    [beamType, phases, stationVersionReferences]
  );
  return useBeamformingTemplates(args);
};

/**
 * Hook to obtain beamforming templates for visible stations and favorite phases for EVENT
 */
export const useBeamformingTemplatesForEvent = (phases?: string[]) => {
  return useBeamformingTemplatesForVisibleStationsAndFavoritePhases(
    BeamformingTemplateTypes.BeamType.EVENT,
    phases
  );
};

/**
 * Hook to obtain beamforming templates for visible stations and favorite phases for FK
 */
export const useBeamformingTemplatesForFK = () => {
  return useBeamformingTemplatesForVisibleStationsAndFavoritePhases(
    BeamformingTemplateTypes.BeamType.FK
  );
};
