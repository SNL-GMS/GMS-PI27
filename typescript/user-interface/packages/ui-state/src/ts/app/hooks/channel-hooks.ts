import type { WaveformTypes } from '@gms/common-model';
import type { TimeRange } from '@gms/common-model/lib/common/types';
import type { VersionReference } from '@gms/common-model/lib/faceted';
import { convertToVersionReference } from '@gms/common-model/lib/faceted';
import { UNFILTERED } from '@gms/common-model/lib/filter/types';
import type { Channel } from '@gms/common-model/lib/station-definitions/channel-definitions/channel-definitions';
import { unwrapResult } from '@reduxjs/toolkit';
import max from 'lodash/max';
import min from 'lodash/min';
import React, { useEffect, useMemo } from 'react';

import type { UiChannelSegment } from '../../types';
import {
  selectBeamedChannels,
  selectEventBeams,
  selectFilteredChannels,
  selectRawChannels,
  selectUiChannelSegments
} from '../api';
import { getChannelsByNamesTimeRange } from '../api/data/channel/get-channels-by-names-timerange';
import { UIStateError } from '../error-handling/ui-state-error';
import { flattenHistory, handleCanceledRequests } from '../query/async-fetch-util';
import { useAppDispatch, useAppSelector } from './react-redux-hooks';
import { useGetSignalDetections } from './signal-detection-hooks';
import { useAllStations } from './station-definition-hooks';
import { useViewableInterval } from './waveform-hooks';

/**
 * Uses a selector to return an array of channels
 *
 * @returns returns an array of channels
 */
export const useChannels = () => {
  const rawChannels = useAppSelector(selectRawChannels);
  const beamedChannels = useAppSelector(selectBeamedChannels);
  return useMemo(() => {
    return [...Object.values(rawChannels), ...Object.values(beamedChannels)];
  }, [rawChannels, beamedChannels]);
};

/**
 * Uses a selector to return an array of raw channels
 *
 * @returns returns an array of channels
 */
export const useRawChannels = () => {
  const rawChannels = useAppSelector(selectRawChannels);
  return useMemo(() => {
    return Object.values(rawChannels);
  }, [rawChannels]);
};

/**
 * Uses a selector to return an array of raw channels and convert them to version references
 *
 * @returns returns an array of channels
 */
export const useRawChannelsVersionReference = () => {
  const rawChannels = useRawChannels();

  return useMemo(() => {
    return Object.values(rawChannels).map(channel => convertToVersionReference(channel, 'name'));
  }, [rawChannels]);
};

/**
 * Uses a selector to return an array of derived channels
 *
 * @returns returns an array of channels
 */
export const useBeamedChannels = () => {
  const beamedChannels = useAppSelector(selectBeamedChannels);
  return useMemo(() => {
    return Object.values(beamedChannels);
  }, [beamedChannels]);
};

/**
 * Uses a selector to return the channel record
 *
 * @returns returns a channel record
 */
export const useUnfilteredChannelsRecord = () => {
  const rawChannels = useAppSelector(selectRawChannels);
  const beamedChannels = useAppSelector(selectBeamedChannels);
  return useMemo(() => {
    return { ...rawChannels, ...beamedChannels };
  }, [rawChannels, beamedChannels]);
};

/**
 * Uses a selector to return all channels from the channel record
 *
 * @returns returns a channel record
 */
export const useAllChannelsRecord = () => {
  const rawChannels = useAppSelector(selectRawChannels);
  const beamedChannels = useAppSelector(selectBeamedChannels);
  const filteredChannels = useAppSelector(selectFilteredChannels);
  return useMemo(() => {
    return { ...rawChannels, ...beamedChannels, ...filteredChannels };
  }, [rawChannels, beamedChannels, filteredChannels]);
};

/**
 * Queries for all channels, without consideration for their visibility
 *
 * Uses the `useGetSignalDetections` hook to get derived channels.
 */
export const useGetChannelsQuery = () => {
  const dispatch = useAppDispatch();
  const [viewableInterval] = useViewableInterval();

  // useGetSignalDetections will populate uiChannelSegments with derived channels
  // so it must be called here to get ALL channels, not just raw
  useGetSignalDetections();

  const history = useAppSelector(state => state.data.queries.getChannelsByNamesTimeRange);

  const allStations = useAllStations();
  const uiChannelSegments = useAppSelector(selectUiChannelSegments);

  const rawChannelNames = useMemo(() => {
    if (!allStations) return [];
    return allStations.flatMap(station => station.allRawChannels.map(channel => channel.name));
  }, [allStations]);

  const derivedChannelNames = useMemo(() => {
    const allChannelsSet = new Set(
      Object.values(uiChannelSegments).flatMap(value => {
        if (!value[UNFILTERED]) return [];
        return value[UNFILTERED].map(
          uiChannelSegment => uiChannelSegment.channelSegment.id.channel.name
        );
      })
    );
    // Unique list of all channel names
    return Array.from(allChannelsSet);
  }, [uiChannelSegments]);

  const eventBeams = useAppSelector(selectEventBeams);
  const eventBeamUiChannelSegments: UiChannelSegment<WaveformTypes.Waveform>[] = useMemo(
    () => Object.values(eventBeams).flatMap(value => value),
    [eventBeams]
  );
  const eventBeamChannelNames = useMemo(() => {
    const allChannelsSet = new Set(
      Object.values(eventBeamUiChannelSegments).map(value => value.channelSegment.id.channel.name)
    );
    // Unique list of all channel names
    return Array.from(allChannelsSet);
  }, [eventBeamUiChannelSegments]);

  // Returns the interval extended to include the channels start / end times
  // This is required because event requests can return channels out of the current viewable interval.
  const extendedInterval = useMemo(() => {
    if (!viewableInterval?.startTimeSecs || !viewableInterval?.endTimeSecs) return viewableInterval;
    const flatUiChannelSegments = Object.values(uiChannelSegments)
      .flatMap(filterRecord => Object.values(filterRecord).flatMap(cs => cs))
      .filter(cs => derivedChannelNames.indexOf(cs.channelSegmentDescriptor.channel.name) >= 0);

    // include the {@link UiChannelSegment}s from the event beams
    flatUiChannelSegments.push(...eventBeamUiChannelSegments);

    return flatUiChannelSegments.reduce(
      (interval, cs) => ({
        ...interval,
        startTimeSecs: min([interval.startTimeSecs, cs.channelSegmentDescriptor.startTime]) || null,
        endTimeSecs: max([interval.endTimeSecs, cs.channelSegmentDescriptor.endTime]) || null
      }),
      { ...viewableInterval }
    );
  }, [uiChannelSegments, derivedChannelNames, eventBeamUiChannelSegments, viewableInterval]);

  // !exclude the derived channel names from the collection; those channels were created with the {@link UiChannelFactory}
  const channelNames = useMemo(
    () => [...rawChannelNames, ...eventBeamChannelNames],
    [rawChannelNames, eventBeamChannelNames]
  );

  //! useEffect updates redux state
  useEffect(() => {
    if (viewableInterval.startTimeSecs != null && viewableInterval.endTimeSecs != null) {
      dispatch(
        getChannelsByNamesTimeRange({
          channelNames,
          startTime: viewableInterval.startTimeSecs,
          endTime: viewableInterval.endTimeSecs
        })
      ).catch(error => {
        throw new UIStateError(error);
      });
    }
  }, [dispatch, viewableInterval.startTimeSecs, viewableInterval.endTimeSecs, channelNames]);

  //! useEffect updates redux state
  useEffect(() => {
    const entries = flattenHistory(history);
    if (extendedInterval?.startTimeSecs != null && extendedInterval?.endTimeSecs != null) {
      const cachedChannelNames = entries
        .filter(
          e =>
            extendedInterval.startTimeSecs != null &&
            extendedInterval.endTimeSecs != null &&
            e.arg.startTime >= extendedInterval.startTimeSecs &&
            e.arg.endTime <= extendedInterval.endTimeSecs
        )
        .flatMap(e => e.arg.channelNames);
      const filteredDerivedChannelNames = derivedChannelNames.filter(
        name => cachedChannelNames.indexOf(name) < 0
      );

      dispatch(
        getChannelsByNamesTimeRange({
          channelNames: filteredDerivedChannelNames,
          startTime: extendedInterval.startTimeSecs,
          endTime: extendedInterval.endTimeSecs
        })
      ).catch(error => {
        throw new UIStateError(error);
      });
    }
  }, [
    derivedChannelNames,
    dispatch,
    extendedInterval?.startTimeSecs,
    extendedInterval?.endTimeSecs,
    history
  ]);
};

/**
 * @returns a callback function which may be used to fetch raw channels
 * The callback takes two args, an array of channel version references,
 * and a timeRange corresponding to the time range for which to fetch
 * channel versions (possibly returning multiple channels if the channel changed
 * during that time period)
 */
export function useFetchRawChannels() {
  const dispatch = useAppDispatch();
  return React.useCallback(
    async (
      channels: VersionReference<'name', Channel>[],
      timeRange: TimeRange
    ): Promise<Channel[]> => {
      return dispatch(
        getChannelsByNamesTimeRange({
          channelNames: channels.map(channel => channel.name),
          startTime: timeRange.startTimeSecs,
          endTime: timeRange.endTimeSecs
        })
      )
        .then(handleCanceledRequests(unwrapResult))
        .catch(error => {
          throw new UIStateError(error);
        });
    },
    [dispatch]
  );
}
