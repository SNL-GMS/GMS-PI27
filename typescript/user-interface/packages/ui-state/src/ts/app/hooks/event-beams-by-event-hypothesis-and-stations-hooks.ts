import type { EventTypes, FacetedTypes } from '@gms/common-model';
import { findPreferredEventHypothesisByOpenStageOrDefaultStage } from '@gms/common-model/lib/event';
import React from 'react';

import type {
  FindEventBeamsByEventHypothesisAndStationsQueryArgs,
  UiChannelSegmentByEventHypothesisId
} from '../api';
import {
  findEventBeamsByEventHypothesisAndStations,
  findEventBeamsByEventHypothesisAndStationsQuery,
  selectEventBeams,
  selectEvents
} from '../api';
import { UIStateError } from '../error-handling/ui-state-error';
import type { AsyncFetchResult, FetchHistoryStatus } from '../query';
import { selectOpenIntervalName } from '../state';
import { useFetchHistoryStatus } from './fetch-history-hooks';
import { useAppDispatch, useAppSelector } from './react-redux-hooks';
import { useVisibleStations } from './station-definition-hooks';

/**
 * Defines async fetch result for the channel segments. It contains flags indicating
 * the status of the request.
 *
 * @see {@link AsyncFetchResult}
 */
export type FindEventBeamsByEventHypothesisAndStationsFetchResult =
  AsyncFetchResult<UiChannelSegmentByEventHypothesisId>;

/**
 * A hook that issues the requests for the find event beams by event
 * hypothesis and stations query.
 *
 * @param args the fetch event beams by event hypothesis and stations query args
 */
export const useFetchFindEventBeamsByEventHypothesisAndStationsQuery = (
  args: FindEventBeamsByEventHypothesisAndStationsQueryArgs
): void => {
  const dispatch = useAppDispatch();

  React.useEffect(() => {
    dispatch(findEventBeamsByEventHypothesisAndStations(args)).catch(error => {
      throw new UIStateError(error);
    });
  }, [dispatch, args]);
};

const useFindEventBeamsByEventHypothesisAndStationsHistory = (): FetchHistoryStatus => {
  const history = useAppSelector(
    state => state.data.queries.findEventBeamsByEventHypothesisAndStations
  );
  return useFetchHistoryStatus<FindEventBeamsByEventHypothesisAndStationsQueryArgs>(history);
};

/**
 * @returns the skipped result for the get channel segments by channels query
 */
const useFindEventBeamsByEventHypothesisAndStationsSkippedResult =
  (): FindEventBeamsByEventHypothesisAndStationsFetchResult => {
    const result = React.useRef({
      data: {},
      pending: 0,
      fulfilled: 0,
      rejected: 0,
      isLoading: false,
      isError: false
    });
    return result.current;
  };

/**
 * A hook that can be used to retrieve channel segments by channels.
 * Makes an individual async request for each channel.
 *
 * This includes the following information:
 *  - the async fetch status of all the async requests
 *  - the `data`: the channel segments from all requests
 *
 * ! the returned results are filtered so that the results only match what the query args requested
 *
 * @param args the channel segments by channels query arguments
 *
 * @returns the channel segments fetch result.
 */
export const useFindEventBeamsByEventHypothesisAndStationsQuery = (
  args: FindEventBeamsByEventHypothesisAndStationsQueryArgs
): FindEventBeamsByEventHypothesisAndStationsFetchResult => {
  const history = useFindEventBeamsByEventHypothesisAndStationsHistory();

  // issue any new fetch requests
  useFetchFindEventBeamsByEventHypothesisAndStationsQuery(args);

  // retrieve all channel segments from the state
  const eventBeams = useAppSelector(selectEventBeams);
  const skippedReturnValue = useFindEventBeamsByEventHypothesisAndStationsSkippedResult();

  return React.useMemo(() => {
    if (findEventBeamsByEventHypothesisAndStationsQuery.shouldSkip(args)) {
      return skippedReturnValue;
    }
    return { ...history, data: eventBeams };
  }, [args, eventBeams, history, skippedReturnValue]);
};

// hook to get everything

export const useFindEventBeamsByEventHypothesisAndStationsQueryForOpenEvent =
  (): FindEventBeamsByEventHypothesisAndStationsFetchResult => {
    const visibleStations = useVisibleStations();
    const events = useAppSelector(selectEvents);
    const openIntervalName = useAppSelector(selectOpenIntervalName);
    const args: FindEventBeamsByEventHypothesisAndStationsQueryArgs = React.useMemo(() => {
      const stations = visibleStations
        ? visibleStations.map(station => ({ name: station.name }))
        : [];
      const preferredHypotheses: FacetedTypes.EntityReference<'id', EventTypes.EventHypothesis>[] =
        [];
      Object.values(events).forEach(event => {
        const eventHypothesis = findPreferredEventHypothesisByOpenStageOrDefaultStage(
          event,
          openIntervalName
        );
        if (eventHypothesis?.id) {
          preferredHypotheses.push({ id: eventHypothesis.id });
        }
      });
      return { eventHypotheses: preferredHypotheses, stations };
    }, [events, openIntervalName, visibleStations]);

    return useFindEventBeamsByEventHypothesisAndStationsQuery(args);
  };
