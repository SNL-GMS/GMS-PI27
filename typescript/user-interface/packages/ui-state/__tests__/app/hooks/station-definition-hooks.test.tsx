/* eslint-disable @typescript-eslint/no-magic-numbers */
import type { ConfigurationTypes } from '@gms/common-model';
import { StationTypes } from '@gms/common-model';
import React from 'react';
import { Provider } from 'react-redux';
import { create } from 'react-test-renderer';

import { useGetAllStationsQuery } from '../../../src/ts/app/hooks/station-definition-hooks';
import { getStore } from '../../../src/ts/app/store';

const defaultMockStationGroup: StationTypes.StationGroup[] = [
  {
    description: 'test group',
    effectiveAt: 123,
    effectiveUntil: 456,
    name: 'test group name',
    stations: [
      {
        name: 'station name',
        description: 'station description',
        type: StationTypes.StationType.HYDROACOUSTIC,
        effectiveAt: 123,
        effectiveUntil: 456,
        relativePositionsByChannel: {},
        location: { depthKm: 3, elevationKm: 3, latitudeDegrees: 1, longitudeDegrees: 3 },
        allRawChannels: [],
        channelGroups: []
      }
    ]
  }
];

const defaultMockStation: StationTypes.Station[] = [
  {
    name: 'station name',
    description: 'station description',
    type: StationTypes.StationType.HYDROACOUSTIC,
    effectiveAt: 123,
    effectiveUntil: 456,
    relativePositionsByChannel: {},
    location: { depthKm: 3, elevationKm: 3, latitudeDegrees: 1, longitudeDegrees: 3 },
    allRawChannels: [],
    channelGroups: []
  }
];

const defaultMockAnalystConfiguration: Partial<ConfigurationTypes.ProcessingAnalystConfiguration> =
  {
    defaultInteractiveAnalysisStationGroup: 'test'
  };

const defaultMockStationGroupNamesConfiguration: Partial<ConfigurationTypes.StationGroupNamesConfiguration> =
  {
    stationGroupNames: ['test']
  };

let mockStationGroup = defaultMockStationGroup;
let mockStation = defaultMockStation;
let mockAnalystConfiguration = defaultMockAnalystConfiguration;
let mockStationGroupNamesConfiguration = defaultMockStationGroupNamesConfiguration;

jest.mock(
  '../../../src/ts/app/api/processing-configuration/processing-configuration-api-slice',
  () => {
    const actual = jest.requireActual(
      '../../../src/ts/app/api/processing-configuration/processing-configuration-api-slice'
    );
    return {
      ...actual,
      useGetProcessingAnalystConfigurationQuery: jest.fn(() => ({
        data: mockAnalystConfiguration
      })),
      useGetProcessingStationGroupNamesConfigurationQuery: jest.fn(() => ({
        data: mockStationGroupNamesConfiguration
      }))
    };
  }
);

jest.mock('../../../src/ts/app/api/station-definition/station-definition-api-slice', () => {
  const actual = jest.requireActual(
    '../../../src/ts/app/api/station-definition/station-definition-api-slice'
  );
  return {
    ...actual,
    useGetStationGroupsByNamesQuery: jest.fn(() => ({
      data: mockStationGroup
    })),
    useGetStationsQuery: jest.fn(() => ({
      data: mockStation
    })),
    useGetStationsWithChannelsQuery: jest.fn(() => ({
      data: mockStation
    }))
  };
});

describe('station definition hooks', () => {
  it('exists', () => {
    expect(useGetAllStationsQuery).toBeDefined();
  });

  it('can use get all stations query', () => {
    mockStationGroup = defaultMockStationGroup;
    mockStation = defaultMockStation;
    mockAnalystConfiguration = defaultMockAnalystConfiguration;
    mockStationGroupNamesConfiguration = defaultMockStationGroupNamesConfiguration;

    const store = getStore();

    function Component() {
      const result = useGetAllStationsQuery(123456789);
      return <div>{JSON.stringify(result.data)}</div>;
    }

    expect(
      create(
        <Provider store={store}>
          <Component />
        </Provider>
      ).toJSON()
    ).toMatchSnapshot();
  });
});
