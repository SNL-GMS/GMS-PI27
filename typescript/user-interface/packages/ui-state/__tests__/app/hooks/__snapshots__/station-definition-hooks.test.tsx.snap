// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`station definition hooks can use get all stations query 1`] = `
<div>
  [{"name":"station name","description":"station description","type":"HYDROACOUSTIC","effectiveAt":123,"effectiveUntil":456,"relativePositionsByChannel":{},"location":{"depthKm":3,"elevationKm":3,"latitudeDegrees":1,"longitudeDegrees":3},"allRawChannels":[],"channelGroups":[]}]
</div>
`;
